run.cc

```diff
@@ -67,13 +68,91 @@ int main() {
   runTests();
   cout << "******************************************************" << endl;
   // create an object which holds data of experiment A
-  Data datA("exp_A");
+  vector<Data> data;
+  Data datA("exp_A"); data.push_back(datA);
+  Data datB("exp_B"); data.push_back(datB);
+  Data datC("exp_C"); data.push_back(datC);
+  Data datD("exp_D"); data.push_back(datD);
 
-  // here is the data from experiment A
-  cout << "bin 27: from " << datA.binLow(27) << " to " << datA.binHigh(27)
+  std::vector<Data> data1 = {datA, datB, datC, datD};
+
+  vector<string> data_labels = {"A", "B", "C", "D"};
+  // here is the data from all experiments
+  for (int i=0; i < 4; ++i){
+    cout << "bin 27: from  experiment " << data_labels[i] << ": " << data[i].binLow(27) << " to " <<  data[i].binHigh(27)
+       << endl;
+  cout << "measurement of experiment " << i << " in bin 27: " <<  data[i].measurement(27)
+       << endl;
+  }
+
+  int deviation_counter = 0;
+  for (int i = 0; i < data[1].size(); ++i){
+    double y1 = data[1].measurement(i), sigmay1 = data[1].error(i);
+    double y2 = data[2].measurement(i), sigmay2 = data[2].error(i);
+    
+    double dy = y1 - y2;
+    double sigmay = sqrt(pow(sigmay1, 2) + pow(sigmay2, 2));
+  
+    if (abs(dy) > sigmay){
+        ++deviation_counter;
+      }
+    for (int n=1; n < 2; ++n){
+      //cout << n << "sigma: " << abs(dy) << " < " << n * sigmay << "?" << endl;
+    }
+  }
+  cout << "Deviating bins: " << deviation_counter << endl;
+  
+  
+  /*
+  cout << data[0].checkCompatibility(data[0], 1) << endl;
+  cout << data[1].checkCompatibility(data[1], 1) << endl;
+  cout << data[2].checkCompatibility(data[2], 1) << endl;
+  cout << data[3].checkCompatibility(data[3], 1) << endl;
+
+  cout << data[0].checkCompatibility(data[1], 1) << endl;
+  cout << data[0].checkCompatibility(data[2], 1) << endl;
+  cout << data[0].checkCompatibility(data[3], 1) << endl;
+  
+  cout << data[1].checkCompatibility(data[2], 1) << endl;
+  cout << data[1].checkCompatibility(data[3], 1) << endl;
+
+  cout << data[2].checkCompatibility(data[3], 1) << endl;
+  */
+  cout << "*************  2  a    **************" << endl;
+
+  for (int n = 1; n < 4; ++n){
+    cout << "Differ by more than " << n << " standard deviations:" << endl;
+    for (int i = 0; i < data1.size()-1; ++i){
+      cout << "*************" << endl;
+      cout << "Comparing " << data_labels[i] << " to ..." << endl;
+      for (int j = i+1; j < data1.size(); ++j){
+        cout << "... " << data_labels[j] << ": "<< data1[i].checkCompatibility(data1[j], n) << endl;
+      }
+      cout << "*************" << endl;
+    }
+  }
+  
```

Comment:
In 2a) you didn't answer how many deviations you would typically expect even if the data were compatible. If you draw two samples from the same distribution, you would still expect some variation and even the probability to see a datapoint that has a deviation of 3 sigma is non-zero. So what you should have done here is just plug in the gaussian probability outside of 3 sigma (and 2 sigma) and multiply this with the number of datapoints, then see if the number of actual deviations agrees with what you computed.  Since your print out is correct, you'll still get 0.5 points for 2a)